File: Sources/ImageHub/Services/PayloadBuilder.swift
import Foundation

/// Writes the `ImageHub\` folder that rides along on the USB drive and is staged
/// to `C:\ImageHub` during Setup: the provisioning scripts, a resolved
/// `config.json`, bundled installers, and any assets the template references.
enum PayloadBuilder {
    static let folderName = "ImageHub"
    static let configFileName = "config.json"

    struct PayloadError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    // MARK: - Config written next to the scripts

    /// Mirror of `Shared/schema/config.schema.json`. `Provision.ps1` reads exactly
    /// these keys, so any change here needs the PowerShell side updated too.
    struct Config: Encodable {
        struct Admin: Encodable {
            var username: String
            var displayName: String
            var accountDescription: String
            var hideFromLoginScreen: Bool
            var passwordNeverExpires: Bool
        }

        struct EndUser: Encodable {
            var mode: String
            var username: String
            var displayName: String
            var administrator: Bool
            var mustChangePassword: Bool
            var welcomeNote: String
        }

        struct Identity: Encodable {
            var joinMode: String
            var workgroup: String
            var domain: String
        }

        struct App: Encodable {
            var name: String
            var source: String
            var packageID: String
            var version: String
            var installer: String
            var silentArgs: String
            var script: String
            var required: Bool
        }

        struct Wifi: Encodable {
            var enabled: Bool
            var ssid: String
            var password: String
            var security: String
            var hidden: Bool
            var connectAutomatically: Bool
        }

        struct Registry: Encodable {
            var path: String
            var name: String
            var type: String
            var value: String
        }

        struct System: Encodable {
            var computerNameTemplate: String
            var timeZone: String
            var enableRemoteDesktop: Bool
            var allowPing: Bool
            var powerPlanGUID: String
            var disableSleepOnAC: Bool
            var disableFastStartup: Bool
            var disableHibernation: Bool
            var showFileExtensions: Bool
            var showHiddenFiles: Bool
            var classicContextMenu: Bool
            var taskbarAlignLeft: Bool
            var disableWidgets: Bool
            var disableWebSearch: Bool
            var disableTelemetry: Bool
            var disableConsumerFeatures: Bool
            var removeBloatware: Bool
            var bloatware: [String]
            var windowsUpdate: String
            var installUpdates: Bool
            var optionalFeatures: [String]
            var bitLocker: String
            var bitLockerRecoveryToAD: Bool
            var disableRecoveryEnvironment: Bool
            var wallpaper: String
            var lockScreen: String
            var startLayout: String
            var wifi: Wifi
            var registryTweaks: [Registry]
        }

        struct Script: Encodable {
            var name: String
            var phase: String
            var file: String
            var continueOnError: Bool
        }

        var schemaVersion = 1
        var generatedBy: String
        var generatedAt: String
        var templateID: String
        var templateName: String
        var admin: Admin
        var endUser: EndUser
        var identity: Identity
        var apps: [App]
        var system: System
        var scripts: [Script]
    }

    // MARK: - Build

    /// Assembles the payload at `<volume>/ImageHub`.
    @discardableResult
    static func write(
        template: DeploymentTemplate,
        secrets: AnswerFileBuilder.ResolvedSecrets,
        to volume: URL,
        log: @escaping @Sendable (String) -> Void
    ) throws -> URL {
        let fm = FileManager.default
        let root = volume.appendingPathComponent(folderName, isDirectory: true)

        if fm.fileExists(atPath: root.path) {
            try fm.removeItem(at: root)
        }
        try fm.createDirectory(at: root, withIntermediateDirectories: true)

        // 1. Provisioning scripts, shared verbatim with the Windows-side builder.
        let source = try payloadSourceDirectory()
        log("Copying provisioning scripts from \(source.lastPathComponent)…")
        for item in try fm.contentsOfDirectory(at: source, includingPropertiesForKeys: nil) {
            guard item.lastPathComponent != ".DS_Store" else { continue }
            try fm.copyItem(at: item, to: root.appendingPathComponent(item.lastPathComponent))
        }

        // 2. Bundled installers.
        var apps: [Config.App] = []
        let installers = root.appendingPathComponent("Installers", isDirectory: true)
        for app in template.enabledApps {
            var relativeInstaller = ""
            if app.source == .installer {
                let origin = URL(fileURLWithPath: app.installerPath)
                guard fm.fileExists(atPath: origin.path) else {
                    throw PayloadError(
                        message: "Installer for “\(app.displayName)” is missing: \(app.installerPath)"
                    )
                }
                if !fm.fileExists(atPath: installers.path) {
                    try fm.createDirectory(at: installers, withIntermediateDirectories: true)
                }
                let destination = installers.appendingPathComponent(origin.lastPathComponent)
                if !fm.fileExists(atPath: destination.path) {
                    log("Copying installer \(origin.lastPathComponent)…")
                    try fm.copyItem(at: origin, to: destination)
                }
                relativeInstaller = "Installers\\\(origin.lastPathComponent)"
            }
            apps.append(
                Config.App(
                    name: app.displayName,
                    source: app.source.rawValue,
                    packageID: app.packageID,
                    version: app.version,
                    installer: relativeInstaller,
                    silentArgs: app.silentArgs,
                    script: app.script,
                    required: app.required
                )
            )
        }

        // 3. Assets referenced by the template.
        let assets = root.appendingPathComponent("Assets", isDirectory: true)
        func copyAsset(_ path: String, as name: String) throws -> String {
            guard !path.isEmpty else { return "" }
            let origin = URL(fileURLWithPath: path)
            guard fm.fileExists(atPath: origin.path) else {
                log("⚠︎ Asset not found, skipping: \(path)")
                return ""
            }
            if !fm.fileExists(atPath: assets.path) {
                try fm.createDirectory(at: assets, withIntermediateDirectories: true)
            }
            let ext = origin.pathExtension
            let fileName = ext.isEmpty ? name : "\(name).\(ext)"
            try fm.copyItem(at: origin, to: assets.appendingPathComponent(fileName))
            return "Assets\\\(fileName)"
        }

        let wallpaper = try copyAsset(template.system.wallpaperPath, as: "Wallpaper")
        let lockScreen = try copyAsset(template.system.lockScreenPath, as: "LockScreen")
        let startLayout = try copyAsset(template.system.startLayoutPath, as: "StartLayout")

        // 4. Custom scripts.
        var scripts: [Config.Script] = []
        let scriptsDirectory = root.appendingPathComponent("Scripts", isDirectory: true)
        for script in template.scripts where script.enabled && !script.body.isEmpty {
            if !fm.fileExists(atPath: scriptsDirectory.path) {
                try fm.createDirectory(at: scriptsDirectory, withIntermediateDirectories: true)
            }
            let name = scriptFileName(script)
            try Data(script.body.utf8)
                .write(to: scriptsDirectory.appendingPathComponent(name))
            scripts.append(
                Config.Script(
                    name: script.name,
                    phase: script.phase.rawValue,
                    file: "Scripts\\\(name)",
                    continueOnError: script.continueOnError
                )
            )
        }

        // 5. Resolved config + an audit copy of the template itself.
        let config = makeConfig(
            template: template,
            secrets: secrets,
            apps: apps,
            scripts: scripts,
            wallpaper: wallpaper,
            lockScreen: lockScreen,
            startLayout: startLayout
        )

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(config).write(to: root.appendingPathComponent(configFileName))

        let templateEncoder = JSONEncoder()
        templateEncoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        templateEncoder.dateEncodingStrategy = .iso8601
        try templateEncoder.encode(template)
            .write(to: root.appendingPathComponent("template.json"))

        log("Payload written to \(root.path)")
        return root
    }

    static func scriptFileName(_ script: CustomScript) -> String {
        let allowed = Set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_")
        let slug = String(script.name.compactMap { allowed.contains($0) ? $0 : "-" }.prefix(40))
        return "\(slug.isEmpty ? "script" : slug)-\(script.id.uuidString.prefix(8)).ps1"
    }

    private static func makeConfig(
        template: DeploymentTemplate,
        secrets: AnswerFileBuilder.ResolvedSecrets,
        apps: [Config.App],
        scripts: [Config.Script],
        wallpaper: String,
        lockScreen: String,
        startLayout: String
    ) -> Config {
        let system = template.system
        return Config(
            generatedBy: "ImageHub \(AppVersion.current)",
            generatedAt: ISO8601DateFormatter().string(from: Date()),
            templateID: template.id.uuidString,
            templateName: template.name,
            admin: Config.Admin(
                username: template.admin.enabled ? template.admin.username : "",
                displayName: template.admin.displayName,
                accountDescription: template.admin.accountDescription,
                hideFromLoginScreen: template.admin.hideFromLoginScreen,
                passwordNeverExpires: template.admin.passwordNeverExpires
            ),
            endUser: Config.EndUser(
                mode: template.endUser.mode.rawValue,
                username: template.endUser.username,
                displayName: template.endUser.displayName,
                administrator: template.endUser.administrator,
                mustChangePassword: template.endUser.mustChangePassword,
                welcomeNote: template.endUser.welcomeNote
            ),
            identity: Config.Identity(
                joinMode: template.identity.joinMode.rawValue,
                workgroup: template.identity.workgroup,
                domain: template.identity.domain
            ),
            apps: apps,
            system: Config.System(
                computerNameTemplate: system.computerNameTemplate,
                timeZone: system.timeZone,
                enableRemoteDesktop: system.enableRemoteDesktop,
                allowPing: system.allowPing,
                powerPlanGUID: system.powerPlan.guid,
                disableSleepOnAC: system.disableSleepOnAC,
                disableFastStartup: system.disableFastStartup,
                disableHibernation: system.disableHibernation,
                showFileExtensions: system.showFileExtensions,
                showHiddenFiles: system.showHiddenFiles,
                classicContextMenu: system.classicContextMenu,
                taskbarAlignLeft: system.taskbarAlignLeft,
                disableWidgets: system.disableWidgets,
                disableWebSearch: system.disableWebSearch,
                disableTelemetry: system.disableTelemetry,
                disableConsumerFeatures: system.disableConsumerFeatures,
                removeBloatware: system.removeBloatware,
                bloatware: system.removeBloatware ? system.bloatwareList : [],
                windowsUpdate: system.windowsUpdate.rawValue,
                installUpdates: system.installUpdatesDuringProvisioning,
                optionalFeatures: system.optionalFeatures,
                bitLocker: system.bitLocker.rawValue,
                bitLockerRecoveryToAD: system.enableBitLockerRecoveryToAD,
                disableRecoveryEnvironment: !template.disk.recoveryPartition,
                wallpaper: wallpaper,
                lockScreen: lockScreen,
                startLayout: startLayout,
                wifi: Config.Wifi(
                    enabled: system.wifi.enabled,
                    ssid: system.wifi.ssid,
                    password: system.wifi.enabled ? secrets.wifiPassword : "",
                    security: system.wifi.security,
                    hidden: system.wifi.hidden,
                    connectAutomatically: system.wifi.connectAutomatically
                ),
                registryTweaks: system.registryTweaks
                    .filter { $0.enabled && !$0.name.isEmpty }
                    .map {
                        Config.Registry(path: $0.path, name: $0.name, type: $0.type.rawValue, value: $0.value)
                    }
            ),
            scripts: scripts
        )
    }

    // MARK: - Locating the shared payload

    static let payloadPathOverrideKey = "payloadSourcePath"

    /// The provisioning scripts live in `Shared/payload/` in the repository and
    /// are copied into `ImageHub.app/Contents/Resources/payload` at build time.
    /// Development runs (`swift run`) fall back to the checkout.
    static func payloadSourceDirectory() throws -> URL {
        let fm = FileManager.default
        var candidates: [URL] = []

        if let override = UserDefaults.standard.string(forKey: payloadPathOverrideKey), !override.isEmpty {
            candidates.append(URL(fileURLWithPath: override, isDirectory: true))
        }
        if let bundled = Bundle.main.resourceURL?.appendingPathComponent("payload", isDirectory: true) {
            candidates.append(bundled)
        }

        // Walk up from the executable (…/.build/<config>/ImageHub) looking for the checkout.
        var directory = Bundle.main.bundleURL
        for _ in 0..<8 {
            candidates.append(
                directory.appendingPathComponent("Shared/payload", isDirectory: true)
            )
            directory = directory.deletingLastPathComponent()
        }
        candidates.append(
            URL(fileURLWithPath: fm.currentDirectoryPath, isDirectory: true)
                .appendingPathComponent("Shared/payload", isDirectory: true)
        )

        for candidate in candidates {
            if fm.fileExists(atPath: candidate.appendingPathComponent("Provision.ps1").path) {
                return candidate
            }
        }

        throw PayloadError(
            message: """
                Couldn't find the provisioning scripts (Shared/payload/Provision.ps1). \
                If you're running a development build, launch it from the repository \
                checkout, or set the payload path in Settings → Tools.
                """
        )
    }
}

File: Windows/App/Services/PayloadBuilder.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Text.Json.Serialization;
using ImageHub.Models;
using ImageHub.Support;

namespace ImageHub.Services;

/// <summary>
/// Writes the ImageHub\ folder that rides along on the USB drive and is staged to
/// C:\ImageHub during Setup: the provisioning scripts, a resolved config.json,
/// bundled installers, and any assets the template references.
///
/// A port of Sources/ImageHub/Services/PayloadBuilder.swift. The config is what
/// Provision.ps1 parses, so the key names here are the contract with
/// Shared/payload/Provision.ps1 as much as with the macOS app — any change needs
/// all three updated together.
/// </summary>
public static class PayloadBuilder
{
    public const string FolderName = "ImageHub";

    public const string ConfigFileName = "config.json";

    // MARK: - Config written next to the scripts

    public sealed class PayloadConfig
    {
        public int SchemaVersion { get; set; } = 1;

        public string GeneratedBy { get; set; } = string.Empty;

        public string GeneratedAt { get; set; } = string.Empty;

        [JsonPropertyName("templateID")]
        public string TemplateId { get; set; } = string.Empty;

        public string TemplateName { get; set; } = string.Empty;

        public AdminConfig Admin { get; set; } = new();

        public EndUserConfig EndUser { get; set; } = new();

        public IdentityConfig Identity { get; set; } = new();

        public List<AppConfig> Apps { get; set; } = new();

        public Microsoft365Config Microsoft365 { get; set; } = new();

        public SystemConfig System { get; set; } = new();

        public List<ScriptConfig> Scripts { get; set; } = new();
    }

    public sealed class AdminConfig
    {
        public string Username { get; set; } = string.Empty;

        public string DisplayName { get; set; } = string.Empty;

        public string AccountDescription { get; set; } = string.Empty;

        public bool HideFromLoginScreen { get; set; }

        public bool PasswordNeverExpires { get; set; }
    }

    public sealed class EndUserConfig
    {
        public string Mode { get; set; } = string.Empty;

        public string Username { get; set; } = string.Empty;

        public string DisplayName { get; set; } = string.Empty;

        public bool Administrator { get; set; }

        public bool MustChangePassword { get; set; }

        public string WelcomeNote { get; set; } = string.Empty;

        public int PromptTimeoutMinutes { get; set; }
    }

    public sealed class IdentityConfig
    {
        public string JoinMode { get; set; } = string.Empty;

        public string Workgroup { get; set; } = string.Empty;

        public string Domain { get; set; } = string.Empty;
    }

    public sealed class AppConfig
    {
        public string Name { get; set; } = string.Empty;

        public string Source { get; set; } = string.Empty;

        [JsonPropertyName("packageID")]
        public string PackageId { get; set; } = string.Empty;

        public string Version { get; set; } = string.Empty;

        public string Installer { get; set; } = string.Empty;

        public string SilentArgs { get; set; } = string.Empty;

        public string Script { get; set; } = string.Empty;

        public bool Required { get; set; }
    }

    public sealed class WifiConfig
    {
        public bool Enabled { get; set; }

        [JsonPropertyName("ssid")]
        public string Ssid { get; set; } = string.Empty;

        public string Password { get; set; } = string.Empty;

        public string Security { get; set; } = string.Empty;

        public bool Hidden { get; set; }

        public bool ConnectAutomatically { get; set; }
    }

    public sealed class Microsoft365Config
    {
        public bool Enabled { get; set; }

        public string Setup { get; set; } = string.Empty;

        public string Configuration { get; set; } = string.Empty;

        public int TimeoutMinutes { get; set; }
    }

    public sealed class ActivationConfig
    {
        public string Mode { get; set; } = string.Empty;

        [JsonPropertyName("kmsHost")]
        public string KmsHost { get; set; } = string.Empty;
    }

    public sealed class RegistryConfig
    {
        public string Path { get; set; } = string.Empty;

        public string Name { get; set; } = string.Empty;

        public string Type { get; set; } = string.Empty;

        public string Value { get; set; } = string.Empty;
    }

    public sealed class SystemConfig
    {
        public string ComputerNameTemplate { get; set; } = string.Empty;

        public string TimeZone { get; set; } = string.Empty;

        public bool EnableRemoteDesktop { get; set; }

        public bool AllowPing { get; set; }

        [JsonPropertyName("powerPlanGUID")]
        public string PowerPlanGuid { get; set; } = string.Empty;

        [JsonPropertyName("disableSleepOnAC")]
        public bool DisableSleepOnAc { get; set; }

        public bool DisableFastStartup { get; set; }

        public bool DisableHibernation { get; set; }

        // Minutes on this side, seconds in the registry. Provision.ps1 does the
        // conversion, so the config stays readable.
        public int ScreenLockMinutes { get; set; }

        public bool ManagePowerTimeouts { get; set; }

        [JsonPropertyName("displayOffMinutesAC")]
        public int DisplayOffMinutesAc { get; set; }

        [JsonPropertyName("displayOffMinutesDC")]
        public int DisplayOffMinutesDc { get; set; }

        [JsonPropertyName("sleepMinutesAC")]
        public int SleepMinutesAc { get; set; }

        [JsonPropertyName("sleepMinutesDC")]
        public int SleepMinutesDc { get; set; }

        // Resolved to the numeric LIDACTION values here so the payload needs no
        // lookup table of its own.
        [JsonPropertyName("lidCloseActionAC")]
        public int LidCloseActionAc { get; set; }

        [JsonPropertyName("lidCloseActionDC")]
        public int LidCloseActionDc { get; set; }

        public bool ShowFileExtensions { get; set; }

        public bool ShowHiddenFiles { get; set; }

        public bool ClassicContextMenu { get; set; }

        public bool TaskbarAlignLeft { get; set; }

        public bool DisableWidgets { get; set; }

        public bool DisableWebSearch { get; set; }

        public bool DisableTelemetry { get; set; }

        public bool DisableConsumerFeatures { get; set; }

        public bool RemoveBloatware { get; set; }

        public List<string> Bloatware { get; set; } = new();

        public string WindowsUpdate { get; set; } = string.Empty;

        public bool InstallUpdates { get; set; }

        public List<string> OptionalFeatures { get; set; } = new();

        public string BitLocker { get; set; } = string.Empty;

        [JsonPropertyName("bitLockerRecoveryToAD")]
        public bool BitLockerRecoveryToAd { get; set; }

        public bool DisableRecoveryEnvironment { get; set; }

        public string Wallpaper { get; set; } = string.Empty;

        public string LockScreen { get; set; } = string.Empty;

        public string StartLayout { get; set; } = string.Empty;

        public string OrganizationName { get; set; } = string.Empty;

        public string Logo { get; set; } = string.Empty;

        public string SupportPhone { get; set; } = string.Empty;

        [JsonPropertyName("supportURL")]
        public string SupportUrl { get; set; } = string.Empty;

        public bool ShowProvisioningScreen { get; set; }

        public WifiConfig Wifi { get; set; } = new();

        public ActivationConfig Activation { get; set; } = new();

        public List<RegistryConfig> RegistryTweaks { get; set; } = new();
    }

    public sealed class ScriptConfig
    {
        public string Name { get; set; } = string.Empty;

        public string Phase { get; set; } = string.Empty;

        public string File { get; set; } = string.Empty;

        public bool ContinueOnError { get; set; }
    }

    // MARK: - Build

    /// <summary>Assembles the payload at &lt;volume&gt;\ImageHub.</summary>
    public static string Write(
        DeploymentTemplate template,
        ResolvedSecrets secrets,
        string volumeRoot,
        Action<string> log)
    {
        string root = Path.Combine(volumeRoot, FolderName);
        if (Directory.Exists(root)) { Directory.Delete(root, recursive: true); }
        Directory.CreateDirectory(root);

        // 1. Provisioning scripts, shared verbatim with the macOS app and the
        //    PowerShell builder.
        PayloadSource.CopyTo(root, log);

        // 2. Bundled installers.
        var apps = new List<AppConfig>();
        string installers = Path.Combine(root, "Installers");
        foreach (AppSelection app in template.EnabledApps)
        {
            string relativeInstaller = string.Empty;
            if (app.Source == AppSource.Installer)
            {
                if (!File.Exists(app.InstallerPath))
                {
                    throw new BuildException(
                        $"Installer for “{app.DisplayName}” is missing: {app.InstallerPath}");
                }
                Directory.CreateDirectory(installers);
                string fileName = Path.GetFileName(app.InstallerPath);
                string destination = Path.Combine(installers, fileName);
                if (!File.Exists(destination))
                {
                    log($"Copying installer {fileName}…");
                    File.Copy(app.InstallerPath, destination);
                }
                relativeInstaller = "Installers\\" + fileName;
            }
            apps.Add(new AppConfig
            {
                Name = app.DisplayName,
                Source = RawEnum.Of(app.Source),
                PackageId = app.PackageId,
                Version = app.Version,
                Installer = relativeInstaller,
                SilentArgs = app.SilentArgs,
                Script = app.Script,
                Required = app.Required,
            });
        }

        // 3. Microsoft 365 through the Office Deployment Tool.
        //
        // setup.exe is resolved by UsbWriter before this runs — downloaded from
        // Microsoft or the operator's pinned copy — and configuration.xml is generated
        // here so a mistake in it surfaces on this PC, not on a bench.
        var office = new Microsoft365Config
        {
            Enabled = false,
            TimeoutMinutes = Microsoft365Spec.TimeoutMinutes,
        };
        if (template.Microsoft365.Enabled)
        {
            if (!File.Exists(template.Microsoft365.SetupPath))
            {
                throw new BuildException(
                    "Microsoft 365 is enabled but the Office Deployment Tool setup.exe is missing: "
                    + template.Microsoft365.SetupPath);
            }
            string officeDirectory = Path.Combine(root, "Office");
            Directory.CreateDirectory(officeDirectory);

            log($"Copying the Office Deployment Tool ({Path.GetFileName(template.Microsoft365.SetupPath)})…");
            File.Copy(template.Microsoft365.SetupPath, Path.Combine(officeDirectory, "setup.exe"), overwrite: true);

            File.WriteAllText(
                Path.Combine(officeDirectory, OfficeConfigBuilder.FileName),
                OfficeConfigBuilder.Xml(template),
                new UTF8Encoding(false));

            office.Enabled = true;
            office.Setup = "Office\\setup.exe";
            office.Configuration = "Office\\" + OfficeConfigBuilder.FileName;
            log($"Wrote {OfficeConfigBuilder.FileName} for Microsoft 365.");
        }

        // 4. Assets referenced by the template.
        string assets = Path.Combine(root, "Assets");
        string CopyAsset(string path, string name)
        {
            if (path.Length == 0) { return string.Empty; }
            if (!File.Exists(path))
            {
                log($"⚠ Asset not found, skipping: {path}");
                return string.Empty;
            }
            Directory.CreateDirectory(assets);
            string extension = Path.GetExtension(path);
            string fileName = extension.Length == 0 ? name : name + extension;
            File.Copy(path, Path.Combine(assets, fileName), overwrite: true);
            return "Assets\\" + fileName;
        }

        string wallpaper = CopyAsset(template.System.WallpaperPath, "Wallpaper");
        string lockScreen = CopyAsset(template.System.LockScreenPath, "LockScreen");
        string startLayout = CopyAsset(template.System.StartLayoutPath, "StartLayout");
        string logo = CopyAsset(template.System.LogoPath, "Logo");

        // 5. Custom scripts.
        var scripts = new List<ScriptConfig>();
        string scriptsDirectory = Path.Combine(root, "Scripts");
        foreach (CustomScript script in template.Scripts)
        {
            if (!script.Enabled || script.Body.Length == 0) { continue; }
            Directory.CreateDirectory(scriptsDirectory);
            string name = ScriptFileName(script);
            // No BOM and the body verbatim, exactly as the macOS side writes it.
            File.WriteAllText(Path.Combine(scriptsDirectory, name), script.Body, new UTF8Encoding(false));
            scripts.Add(new ScriptConfig
            {
                Name = script.Name,
                Phase = RawEnum.Of(script.Phase),
                File = "Scripts\\" + name,
                ContinueOnError = script.ContinueOnError,
            });
        }

        // 6. Resolved config + an audit copy of the template itself.
        PayloadConfig config = MakeConfig(
            template, secrets, apps, office, scripts, wallpaper, lockScreen, startLayout, logo);

        File.WriteAllText(Path.Combine(root, ConfigFileName), Json.Serialize(config), new UTF8Encoding(false));
        File.WriteAllText(Path.Combine(root, "template.json"), Json.Serialize(template), new UTF8Encoding(false));

        log($"Payload written to {root}");
        return root;
    }

    public static string ScriptFileName(CustomScript script)
    {
        const string allowed = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_";
        var slug = new StringBuilder();
        foreach (char c in script.Name)
        {
            slug.Append(allowed.IndexOf(c) >= 0 ? c : '-');
            if (slug.Length == 40) { break; }
        }
        string stem = slug.Length == 0 ? "script" : slug.ToString();
        string shortId = script.Id.ToString("D").ToUpperInvariant().Substring(0, 8);
        return $"{stem}-{shortId}.ps1";
    }

    private static PayloadConfig MakeConfig(
        DeploymentTemplate template,
        ResolvedSecrets secrets,
        List<AppConfig> apps,
        Microsoft365Config office,
        List<ScriptConfig> scripts,
        string wallpaper,
        string lockScreen,
        string startLayout,
        string logo)
    {
        SystemSpec system = template.System;
        return new PayloadConfig
        {
            GeneratedBy = "ImageHub " + AppVersion.Current,
            GeneratedAt = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture),
            TemplateId = template.Id.ToString("D").ToUpperInvariant(),
            TemplateName = template.Name,
            Admin = new AdminConfig
            {
                Username = template.Admin.Enabled ? template.Admin.Username : string.Empty,
                DisplayName = template.Admin.DisplayName,
                AccountDescription = template.Admin.AccountDescription,
                HideFromLoginScreen = template.Admin.HideFromLoginScreen,
                PasswordNeverExpires = template.Admin.PasswordNeverExpires,
            },
            EndUser = new EndUserConfig
            {
                Mode = RawEnum.Of(template.EndUser.Mode),
                Username = template.EndUser.Username,
                DisplayName = template.EndUser.DisplayName,
                Administrator = template.EndUser.Administrator,
                MustChangePassword = template.EndUser.MustChangePassword,
                WelcomeNote = template.EndUser.WelcomeNote,
                PromptTimeoutMinutes = template.EndUser.PromptTimeoutMinutes,
            },
            Identity = new IdentityConfig
            {
                JoinMode = RawEnum.Of(template.Identity.JoinMode),
                Workgroup = template.Identity.Workgroup,
                Domain = template.Identity.Domain,
            },
            Apps = apps,
            Microsoft365 = office,
            System = new SystemConfig
            {
                ComputerNameTemplate = system.ComputerNameTemplate,
                TimeZone = system.TimeZone,
                EnableRemoteDesktop = system.EnableRemoteDesktop,
                AllowPing = system.AllowPing,
                PowerPlanGuid = Labels.Guid(system.PowerPlan),
                DisableSleepOnAc = system.DisableSleepOnAc,
                DisableFastStartup = system.DisableFastStartup,
                DisableHibernation = system.DisableHibernation,
                ScreenLockMinutes = system.ScreenLockMinutes,
                ManagePowerTimeouts = system.ManagePowerTimeouts,
                DisplayOffMinutesAc = system.DisplayOffMinutesAc,
                DisplayOffMinutesDc = system.DisplayOffMinutesDc,
                SleepMinutesAc = system.SleepMinutesAc,
                SleepMinutesDc = system.SleepMinutesDc,
                LidCloseActionAc = Labels.Index(system.LidCloseActionAc),
                LidCloseActionDc = Labels.Index(system.LidCloseActionDc),
                ShowFileExtensions = system.ShowFileExtensions,
                ShowHiddenFiles = system.ShowHiddenFiles,
                ClassicContextMenu = system.ClassicContextMenu,
                TaskbarAlignLeft = system.TaskbarAlignLeft,
                DisableWidgets = system.DisableWidgets,
                DisableWebSearch = system.DisableWebSearch,
                DisableTelemetry = system.DisableTelemetry,
                DisableConsumerFeatures = system.DisableConsumerFeatures,
                RemoveBloatware = system.RemoveBloatware,
                Bloatware = system.RemoveBloatware ? new List<string>(system.BloatwareList) : new List<string>(),
                WindowsUpdate = RawEnum.Of(system.WindowsUpdate),
                InstallUpdates = system.InstallUpdatesDuringProvisioning,
                OptionalFeatures = new List<string>(system.OptionalFeatures),
                BitLocker = RawEnum.Of(system.BitLocker),
                BitLockerRecoveryToAd = system.EnableBitLockerRecoveryToAd,
                DisableRecoveryEnvironment = !template.Disk.RecoveryPartition,
                Wallpaper = wallpaper,
                LockScreen = lockScreen,
                StartLayout = startLayout,
                OrganizationName = system.OrganizationName,
                Logo = logo,
                SupportPhone = system.SupportPhone,
                SupportUrl = system.SupportUrl,
                ShowProvisioningScreen = system.ShowProvisioningScreen,
                Wifi = new WifiConfig
                {
                    Enabled = system.Wifi.Enabled,
                    Ssid = system.Wifi.Ssid,
                    Password = system.Wifi.Enabled ? secrets.WifiPassword : string.Empty,
                    Security = system.Wifi.Security,
                    Hidden = system.Wifi.Hidden,
                    ConnectAutomatically = system.Wifi.ConnectAutomatically,
                },
                Activation = new ActivationConfig
                {
                    Mode = RawEnum.Of(template.Windows.Activation.Mode),
                    KmsHost = template.Windows.Activation.KmsHost,
                },
                RegistryTweaks = system.RegistryTweaks
                    .Where(tweak => tweak.Enabled && tweak.Name.Length > 0)
                    .Select(tweak => new RegistryConfig
                    {
                        Path = tweak.Path,
                        Name = tweak.Name,
                        Type = RawEnum.Of(tweak.Type),
                        Value = tweak.Value,
                    })
                    .ToList(),
            },
            Scripts = scripts,
        };
    }
}

/// <summary>
/// Where the provisioning scripts come from.
///
/// Shared/payload is the canonical copy; it is embedded in ImageHub.exe at build
/// time so a single .exe carries everything it needs to write a drive, the way the
/// macOS app carries them in its bundle. The Settings override points at a
/// checkout's Shared/payload so a change to Provision.ps1 can be tested without
/// rebuilding ImageHub.
///
/// Files are copied byte for byte and never rewritten. Provision.ps1 has to keep
/// its UTF-8 BOM: Windows PowerShell 5.1 decodes a BOM-less file as ANSI, and CI
/// checks the repository copies for exactly that.
/// </summary>
public static class PayloadSource
{
    private const string ResourcePrefix = "payload/";

    /// <summary>A folder holding Provision.ps1, or null when the embedded copy is used.</summary>
    public static string? OverrideDirectory
    {
        get
        {
            string path = Settings.Current.PayloadSourcePath;
            if (path.Length == 0) { return null; }
            return File.Exists(Path.Combine(path, "Provision.ps1")) ? path : null;
        }
    }

    public static string Describe()
    {
        string? over = OverrideDirectory;
        if (over is not null) { return over; }
        int count = Names().Count;
        return count > 0
            ? $"{count} files embedded in ImageHub.exe"
            : "missing — this build has no provisioning scripts";
    }

    public static IReadOnlyList<string> Names()
    {
        var names = new List<string>();
        foreach (string resource in Assembly.GetExecutingAssembly().GetManifestResourceNames())
        {
            if (resource.StartsWith(ResourcePrefix, StringComparison.Ordinal))
            {
                names.Add(resource.Substring(ResourcePrefix.Length));
            }
        }
        names.Sort(StringComparer.OrdinalIgnoreCase);
        return names;
    }

    public static void CopyTo(string destination, Action<string> log)
    {
        string? over = OverrideDirectory;
        if (over is not null)
        {
            log($"Copying provisioning scripts from {over}…");
            foreach (string file in Directory.GetFiles(over))
            {
                string name = Path.GetFileName(file);
                if (name.Equals(".DS_Store", StringComparison.OrdinalIgnoreCase)) { continue; }
                File.Copy(file, Path.Combine(destination, name), overwrite: true);
            }
            return;
        }

        IReadOnlyList<string> names = Names();
        if (names.Count == 0 || !names.Any(n => n.Equals("Provision.ps1", StringComparison.OrdinalIgnoreCase)))
        {
            throw new BuildException(
                "This build of ImageHub has no provisioning scripts embedded (Shared/payload/Provision.ps1). "
                + "Point Options → Advanced at a repository checkout's Shared\\payload folder, or "
                + "reinstall ImageHub.");
        }

        log($"Writing {names.Count} provisioning script(s) from the app…");
        Assembly assembly = Assembly.GetExecutingAssembly();
        foreach (string name in names)
        {
            using Stream? source = assembly.GetManifestResourceStream(ResourcePrefix + name);
            if (source is null) { continue; }
            using FileStream target = File.Create(Path.Combine(destination, name));
            source.CopyTo(target);
        }
    }

    /// <summary>Reads one embedded script, for the Options page's "what shipped" list.</summary>
    public static byte[]? Read(string name)
    {
        using Stream? source = Assembly.GetExecutingAssembly()
            .GetManifestResourceStream(ResourcePrefix + name);
        if (source is null) { return null; }
        using var buffer = new MemoryStream();
        source.CopyTo(buffer);
        return buffer.ToArray();
    }
}
